### ISSUE TITLE
Contract event listener is not firing when running hardhat tests on ethers v6

### ISSUE THREAD

**Issue by lescaudr** [NomicFoundation/hardhat#4044]
### Version of Hardhat

2.15.0

### What happened?

I'm running into the same issue as this one https://github.com/NomicFoundation/hardhat/issues/1692 from 2 years ago with the correct modification in ethers v6 but the contract.on is not triggered.

```
 await new Promise(async (resolve) => {
      contractA.on("TokensMinted", (amount: BigNumber) => {
        // THIS LINE NEVER GETS HIT
        console.log("###########");
        resolve(true);
      });

      const contractTx: ContractTransaction = await contractA.mint(123);
      const contractReceipt: ContractReceipt = await contractTx.wait();

      for (const event of contractReceipt.logs!) {
        console.log(event.fragment.name, event.args[0].toString());
      }

      // Works
      //  contractA.emit("TokensMinted", 123);
    });
```

I'm adding the Promise to ensure to wait until the event is emitted.
From the receipt, the event is properly emitted, then it wait indefinitely because the event is never emitted.

If I emit the event manually using `contractA.emit("TokensMinted", 123);`, then I get the console.log and the promise is resolved.

### Minimal reproduction steps

Repo to reproduce the issue easily: https://github.com/lescaudr/ethers-event-issue
pnpm install
pnpm compile
pnpm test

**Comment by fvictorio** [comment]
Thank you @lescaudr, I can confirm this. It's an important problem, I'll try to get it fixed asap.

**Comment by tevdoradze** [comment]
Same problem, .on is not triggered. (v2.12.3)

**Comment by fvictorio** [comment]
@tevdoradze it should still work if you use ethers v5 (and the associated plugins: `@nomicfoundation/hardhat-toolbox@2`, `@nomiclabs/hardhat-ethers@2` and `@nomicfoundation/hardhat-chai-matchers@1`)

**Comment by fvictorio** [comment]
Released in [hardhat-ethers v3.0.3](https://github.com/NomicFoundation/hardhat/releases/tag/%40nomicfoundation%2Fhardhat-ethers%403.0.3) :tada:

**Comment by alxshogenov** [comment]
> Released in [hardhat-ethers v3.0.3](https://github.com/NomicFoundation/hardhat/releases/tag/%40nomicfoundation%2Fhardhat-ethers%403.0.3) 🎉

It still doesn't work in tests

**Comment by fvictorio** [comment]
@shaggyrec if you have reproduction steps, please open an issue about it.

**Comment by lescaudr** [comment]
I confirm it works perfectly @shaggyrec, see the branch "fixed" on `https://github.com/lescaudr/ethers-event-issue` where it's now using `"hardhat": "^2.16.1"`. instead of `2.15.0`

Thanks @fvictorio for saving me so much time while learning to develop smart contracts where events emission are crucial

**Comment by shaggyrec** [comment]
Yeah. Seems like working for your case and the standard hardhat `Lock.sol` contract (I rewrote the test with events to test events handling).

But for Patric Collins' tutorial updated to the latest versions it doesn't work. 
https://github.com/shaggyrec/hardhat-smartcontract-lottery-ts/blob/main/test/unit/Raffle.test.ts#L163

I have no idea why.

**Comment by lescaudr** [comment]
You are right @shaggyrec, indexed events are not triggered, I created a new issue https://github.com/NomicFoundation/hardhat/issues/4098. Also found that when triggering multiple events, only the first is triggered.

**Comment by tevdoradze** [comment]
if it doesn't work in your tests you are most like not waiting long enough:
- Change the pollingInterval from 5s to 1s:  `yourcontract.provider.pollingInterval = 1;`
- Resolve the test with done:
```
it("test", function (done) {
   ...
      yourcontract.on(
        ...,
        async (...) => {
          ...
          done();
        }
      );
      ...
  });
```

If you work with an upgradable contract you may need to add `// @ts-ignore` above the line where we change the pollingInterval

**Comment by lescaudr** [comment]
@tevdoradze polling interval is not the resolution here in ether v6. The bug has been confirmed, and is in progress

**Comment by Devankit2022** [comment]
update?

**Comment by lescaudr** [comment]
It has been fixed 3 weeks ago https://github.com/NomicFoundation/hardhat/issues/4098#issuecomment-1635587540
